Fedora/RHEL RPM

cathay4t opened the pull request:
* Contains a fix for running nmstatectl in python3.
 * Contains patch for removing pbr dependent.
 * `packaging/make_rpm.sh` to make rpm for RHEL 7/8 and Fedora 28+.
 * Change the integration test to use compiled rpm instead of `pip install`.

Extra notes:
 * The script `make_rpm.sh` is designed to build rpm from git HEAD, it's not designed to release.
 * For RHEL/Centos 7, it requires EPEL repo enabled.
 * ~~RHEL 8 beta is not supported yet as we are missing `python3-pbr` there. 
   I will add patch to remove dependency of pbr in this PR soon.~~

cathay4t commented:
For release to public, I would suggest we release early and release often, every time a new feature was introduced(or a big bug fix applied), we tag a new lease(change y version if API addition, change z version if bug fix). I would do the Fedora/EPEL packaging works.

Normally people don't need a rpm build from working git master branch, they need rpm built from release tag.

cathay4t commented:
Patches are ready for review. Thank you!

tyll commented:
> > Change the integration test to use compiled rpm instead of `pip install`.
> 
> I am not clear why this is needed.
> Please extract it to a different PR with a reasoning for it so we can discuss it there.

This ensures that the RPM building is working to a certain degree. What other test would you propose for this PR instead? Since setup.py is used in building the RPM and pip would use setup.py as well, using the RPM in the integration tests covers both paths.

EdDev commented:
> This ensures that the RPM building is working to a certain degree. What other test would you propose for this PR instead? Since setup.py is used in building the RPM and pip would use setup.py as well, using the RPM in the integration tests covers both paths.

That dependency between rpm >> setup.py was not known to me. I can't say I like it but I do not know if we have any other choice. Still, it has its own reasoning and explanation, it should have its own PR.

Testing the RPM build does not need to be part of the integration tests, and as it is not the reasoning to remove pip support from the integration tests, I guess it should have its own reasoning. Just note that we need to make sure pip also works.

tyll commented:
@cathay4t the other PRs seem to be merged now so this needs to be rebased and the label  removed

cathay4t commented:
@tyll Rebased and label removed.

tyll commented on automation/Dockerfile:
```diff
@@ -13,30 +13,22 @@ RUN (cd /lib/systemd/system/sysinit.target.wants/; for i in *; do [ $i == \
+    yum-config-manager --enable till-nmstate-NetworkManager && \
+    yum -y install NetworkManager
+
+RUN yum-config-manager --add-repo http://people.redhat.com/fge/el7_pytest/fge_el7_pytest.repo && \
```
I would prefer to use COPR instead of the repository in the people page. That way the packages will also be GPG signed.

About the organization name, to use nmstate on COPR I suppose we need a Fedora user with this name that has access to COPR. Not sure, if this is really worth the trouble to maintain an extra user account since these are just internal details.

cathay4t replied:
Yes, it's public accessible.
I will try to find a better place for it.

tyll commented on packaging/nmstate.py2.spec.in:
```diff
@@ -0,0 +1,61 @@
+%setup -q -n %{srcname}-%{version}
+
+%build
+PBR_VERSION="%{version}" %py2_build
```
Is the PBR_VERSION still required?

tyll commented on packaging/nmstate.py3.spec.in:
```diff
@@ -0,0 +1,61 @@
+Group:          System Environment/Libraries
+License:        GPLv2+
+Requires:       NetworkManager-libnm
+Requires:       python3-gobject
```
python3-gobject-base should be enough AFAIK.

cathay4t commented on automation/Dockerfile:
```diff
@@ -13,30 +13,22 @@ RUN (cd /lib/systemd/system/sysinit.target.wants/; for i in *; do [ $i == \
+
+RUN yum-config-manager --add-repo http://people.redhat.com/fge/el7_pytest/fge_el7_pytest.repo && \
+    yum-config-manager --enable fge-el7-pytest-cov && \
+    yum -y install epel-release centos-release-openstack-rocky && \
```
For the new version of pytest(which ships pytest instead of py.test).

cathay4t commented on automation/Dockerfile:
```diff
@@ -13,30 +13,22 @@ RUN (cd /lib/systemd/system/sysinit.target.wants/; for i in *; do [ $i == \
+# Patch with NetworkManager fix for BZ#1642625
+RUN yum-config-manager --add-repo https://copr.fedorainfracloud.org/coprs/till/nmstate-NetworkManager/repo/epel-7/till-nmstate-NetworkManager-epel-7.repo && \
+    yum-config-manager --enable till-nmstate-NetworkManager && \
+    yum -y install NetworkManager
```
Sure.

cathay4t commented on packaging/make_rpm.sh:
```diff
@@ -0,0 +1,57 @@
+    VERSION="$(VERSION)git$(git rev-parse --short HEAD)"
+fi
+
+if [ "CHK$(rpm -E %{?fedora} 2>/dev/null)" != "CHK" ] ||
```
It's a old habit. Will change that also.

EdDev commented on packaging/nmstate.py2.spec.in:
```diff
@@ -0,0 +1,61 @@
+BuildRequires:  python2-pyyaml
+BuildRequires:  python-jsonschema
+BuildRequires:  python-setuptools
+Requires:       python2-%{libname}
```
No, it does not need NM, probably for the same reason libnm does not require it.
One usage scenario is to run nmstate in a container with just libnm and a mounted dbus socket.

tyll replied:
Ok, then it should be a `Recommends`.

cathay4t replied:
Unfortunately, RHEL/Centos 7 does not support that keyword.

cathay4t commented:
V2 patch uploaded.

Changes since V1:
 * RPM SPEC trivial fixes based suggestions.
 * Changed dockerfile to use [nmstate copr repo][1].
 * Local build docker images works well on integration test. (including `yum upgrade -y`, weird!).

[1]: http://copr.fedorainfracloud.org/coprs/nmstate/nmstate-test-env-el/

cathay4t commented on packaging/nmstate.py2.spec.in:
```diff
@@ -0,0 +1,61 @@
+Requires:       python2-pyyaml
+
+%description -n python2-%{libname}
+This package contains the python library for nmstate.
```
Sure.

cathay4t commented on automation/Dockerfile:
```diff
@@ -13,30 +13,25 @@ RUN (cd /lib/systemd/system/sysinit.target.wants/; for i in *; do [ $i == \
-RUN yum -y install iproute NetworkManager epel-release git cairo* && \
-    yum -y install python2-pip gcc python-devel gobject-introspection-devel && \
+# Patch with NetworkManager fix for BZ#1642625
+RUN yum-config-manager --add-repo https://copr.fedorainfracloud.org/coprs/till/nmstate-NetworkManager/repo/epel-7/till-nmstate-NetworkManager-epel-7.repo && \
```
Sure. Will copy it to nmstate repo.

cathay4t commented:
V3 patch uploaded.

Changes since V2:
 * Rephrase rpm summery and description text.
 * Moved the NetworkManager bug fix build to nmstate copr repo.

Currently copr is down and refusing to build NetworkManager, but that does not impact current CI testing. We just need copr repo ready when uploading new docker image.
